src/multidimensional_matrix_operators.jl

```diff
@@ -0,0 +1,198 @@
+- Jan Glaubitz, Simon-Christian Klein, Jan Nordström, Philipp Öffner (2023)
+  Multi-dimensional summation-by-parts operators for general function spaces: Theory and construction
+  Journal of Computational Physics 491, 112370, [DOI: 10.1016/j.jcp.2023.112370](https://doi.org/10.1016/j.jcp.2023.112370).
+"""
+mass_matrix_boundary(D::MultidimensionalMatrixDerivativeOperator, dim::Int) = Diagonal(weights_boundary_scaled(D, dim))
+
+function get_weight_boundary(D::MultidimensionalMatrixDerivativeOperator, i::Int)
+    @unpack weights_boundary, on_boundary = D
+    N, _ = size(D)
+    @boundscheck begin
+        @argcheck 1 <= i <= N
+    end
+    if !on_boundary[i]
+        return zero(eltype(D))
+    end
+    j = sum(view(on_boundary, 1:i))
+    @inbounds ω = weights_boundary[j]
+    return ω
+end
+
+function get_weight_boundary_scaled(D::MultidimensionalMatrixDerivativeOperator, dim::Int, i::Int)
+    @unpack normals, on_boundary = D
+    if !on_boundary[i]
+        return zero(eltype(D))
+    end
+    ω = get_weight_boundary(D, i)
+    j = sum(view(on_boundary, 1:i))
+    ω * normals[j][dim]
+end
+
+function Base.show(io::IO, D::MultidimensionalMatrixDerivativeOperator)
+    if get(io, :compact, false)
+        summary(io, D)
+    else
+        x = grid(D)
+        print(io, ndims(D), "-dimensional matrix-based first-derivative operator {T=", eltype(D), "}")
+        print(io, " on ", length(x), " nodes")
+    end
+end
+
+# TODO: mul! How? Depends on direction
+
+"""
+    tensor_product_operator_2D(D_x, D_y = D_x)
+
+Create a 2D [`MultidimensionalMatrixDerivativeOperator`](@ref) on a square based on a 1D derivative operators `D_x` and `D_y` using a tensor product structure.
+The operator `D_x` is used in the x-direction and `D_y` in the y-direction.
+
+For the construction, see also:
+- Sigrun Ortleb (2021)
+  Numerical Methods for Fluid Flow:
+    High Order SBP Schemes, IMEX Advection-Diffusion Splitting and Positivity Preservation for Production-Destruction-PDEs
+  Habilitation thesis, University of Kassel, [DOI: 10.17170/kobra-202301037274](https://doi.org/10.17170/kobra-202301037274),
+  Chapter 1.2.4.
+- Magnus Svärd, Jan Nordström (2014)
+  Review of summation-by-parts schemes for initial-boundary-value problems
+  Journal of Computational Physics 268, pp. 17-38, [DOI: 10.1016/j.jcp.2014.02.031](https://doi.org/10.1016/j.jcp.2014.02.031).
+"""
+function tensor_product_operator_2D(D_x, D_y = D_x)
+    T = promote_type(eltype(D_x), eltype(D_y))
+    nodes_1D_x = grid(D_x)
+    nodes_1D_y = grid(D_y)
+    N_x = length(nodes_1D_x)
+    N_y = length(nodes_1D_y)
+    nodes = SVector.(vec(nodes_1D_x' .* ones(N_y)), vec(ones(N_x)' .* nodes_1D_y))
+    on_boundary = zeros(Bool, N_x * N_y)
+    on_boundary[1:N_y] .= true # left boundary
+    on_boundary[N_y:N_y:end-N_y + 1] .= true # lower boundary
+    on_boundary[N_y + 1:N_y:end-N_y + 1] .= true # upper boundary
+    on_boundary[end-N_y+1:end] .= true # right boundary
+
+    D_1D_x = sparse(D_x)
```

Comment:
It's a good question whether `sparse` is the best choice in every case - Lobatto operators are dense.

Reply:
Could we make the interface a bit more general for now, allowing to specify that `sparse` should be used to form a matrix operator, but also allowing a potentially lazy implementation based on a Kronecker product implementation similar to https://github.com/trixi-framework/Trixi.jl/blob/cdc083f0cb04ad82447dcc49007095046e5fa5df/src/solvers/dgmulti/types.jl#L333?

Reply:
I also thought about an optimized way to store the Kronecker products (which are very special in our case because one of the matrices in the Kronecker products are identity matrices, which could be utilized), but I was too lazy to implement it myself and I don't plan to spend the time implementing it. I didn't know about the implementation in Trixi.jl. However, as far as I understand it, it is only meant for Kronecker products, where both matrices are the same, i.e. something like `kron(A, A)`, which is not the case here. Is my understanding correct?
However, it might still make sense to allow for a more general interface, which would work better for dense 1D matrices. I am just a bit unsure how this interface should look like.

Reply:
Let's go for the option you have chosen right now.

Reply:
Sounds good to me. If we get another better idea in the future, this can be optimized later on.